File: trevrpc-go/status.go
package trevrpc

import (
	"context"
	"errors"
	"fmt"
)

// Code is a TrevRPC status code.
type Code uint32

const (
	CodeOK                 Code = 0
	CodeCancelled          Code = 1
	CodeUnknown            Code = 2
	CodeInvalidArgument    Code = 3
	CodeDeadlineExceeded   Code = 4
	CodeNotFound           Code = 5
	CodeAlreadyExists      Code = 6
	CodePermissionDenied   Code = 7
	CodeResourceExhausted  Code = 8
	CodeFailedPrecondition Code = 9
	CodeAborted            Code = 10
	CodeOutOfRange         Code = 11
	CodeUnimplemented      Code = 12
	CodeInternal           Code = 13
	CodeUnavailable        Code = 14
	CodeDataLoss           Code = 15
	CodeUnauthenticated    Code = 16
)

// CodeFromUint32 converts a wire-format numeric status code into a Code.
func CodeFromUint32(code uint32) Code {
	switch Code(code) {
	case CodeOK, CodeCancelled, CodeInvalidArgument, CodeDeadlineExceeded, CodeNotFound, CodeAlreadyExists, CodePermissionDenied, CodeResourceExhausted, CodeFailedPrecondition, CodeAborted, CodeOutOfRange, CodeUnimplemented, CodeInternal, CodeUnavailable, CodeDataLoss, CodeUnauthenticated:
		return Code(code)
	default:
		return CodeUnknown
	}
}

// Status is an RPC status with a code, message, and optional metadata.
type Status struct {
	Code     Code
	Message  string
	Metadata Metadata
}

// NewStatus creates a status from a code and message.
func NewStatus(code Code, message string) *Status {
	return &Status{Code: code, Message: message, Metadata: Metadata{}}
}

// WithMetadata returns a copy of the status carrying terminal metadata.
func (s *Status) WithMetadata(metadata Metadata) *Status {
	if s == nil {
		s = OK()
	}

	return &Status{Code: s.Code, Message: s.Message, Metadata: cloneMetadata(metadata)}
}

// OK creates an OK status.
func OK() *Status { return NewStatus(CodeOK, "") }

// Cancelled creates a cancelled status.
func Cancelled(message string) *Status { return NewStatus(CodeCancelled, message) }

// Unknown creates an unknown status.
func Unknown(message string) *Status { return NewStatus(CodeUnknown, message) }

// Internal creates an internal error status.
func Internal(message string) *Status { return NewStatus(CodeInternal, message) }

// InvalidArgument creates an invalid-argument status.
func InvalidArgument(message string) *Status {
	return NewStatus(CodeInvalidArgument, message)
}

// DeadlineExceeded creates a deadline-exceeded status.
func DeadlineExceeded(message string) *Status { return NewStatus(CodeDeadlineExceeded, message) }

// NotFound creates a not-found status.
func NotFound(message string) *Status { return NewStatus(CodeNotFound, message) }

// ResourceExhausted creates a resource-exhausted status.
func ResourceExhausted(message string) *Status {
	return NewStatus(CodeResourceExhausted, message)
}

// Unavailable creates an unavailable status.
func Unavailable(message string) *Status { return NewStatus(CodeUnavailable, message) }

// FailedPrecondition creates a failed-precondition status.
func FailedPrecondition(message string) *Status {
	return NewStatus(CodeFailedPrecondition, message)
}

// Unauthenticated creates an unauthenticated status.
func Unauthenticated(message string) *Status { return NewStatus(CodeUnauthenticated, message) }

// Unimplemented creates an unimplemented status.
func Unimplemented(message string) *Status { return NewStatus(CodeUnimplemented, message) }

// Error returns the status as a human-readable error string.
func (s *Status) Error() string {
	if s == nil {
		return "<nil>"
	}

	if s.Message == "" {
		return s.Code.String()
	}

	return fmt.Sprintf("%s: %s", s.Code, s.Message)
}

// IsOK reports whether the status is OK.
func (s *Status) IsOK() bool {
	return s != nil && s.Code == CodeOK
}

// IntoResponse converts the status and response body into an RPC response.
func (s *Status) IntoResponse(body []byte) *RpcResponse {
	if s == nil || len(s.Metadata) == 0 {
		return s.IntoResponseWithMetadata(body, Metadata{})
	}

	return s.IntoResponseWithMetadata(body, s.Metadata)
}

// IntoResponseWithMetadata converts the status, response body, and metadata into an RPC response.
func (s *Status) IntoResponseWithMetadata(body []byte, metadata Metadata) *RpcResponse {
	if s == nil {
		s = OK()
	}

	return &RpcResponse{
		Status:   uint32(s.Code),
		Message:  s.Message,
		Body:     body,
		Metadata: metadata,
	}
}

// StatusFromResponse builds a status from an RPC response.
func StatusFromResponse(response *RpcResponse) *Status {
	if response == nil {
		return Internal("missing RPC response")
	}

	return NewStatus(CodeFromUint32(response.Status), response.Message).WithMetadata(response.Metadata)
}

// StatusFromError converts an error into the status returned to RPC callers.
func StatusFromError(err error) *Status {
	if err == nil {
		return OK()
	}

	var status *Status
	if errors.As(err, &status) {
		return status
	}

	var frameTooLarge *FrameTooLargeError
	if errors.As(err, &frameTooLarge) {
		return ResourceExhausted(frameTooLarge.Error())
	}

	var frameDecode *FrameDecodeError
	if errors.As(err, &frameDecode) {
		return InvalidArgument(frameDecode.Error())
	}

	return Internal(err.Error())
}

func statusFromContextError(err error) *Status {
	if errors.Is(err, context.DeadlineExceeded) {
		return DeadlineExceeded("RPC deadline exceeded")
	}

	if errors.Is(err, context.Canceled) {
		return Cancelled("RPC cancelled")
	}

	return Unavailable("context unavailable: " + err.Error())
}

// String returns the canonical status code name.
func (c Code) String() string {
	switch c {
	case CodeOK:
		return "Ok"
	case CodeCancelled:
		return "Cancelled"
	case CodeInvalidArgument:
		return "InvalidArgument"
	case CodeDeadlineExceeded:
		return "DeadlineExceeded"
	case CodeNotFound:
		return "NotFound"
	case CodeAlreadyExists:
		return "AlreadyExists"
	case CodePermissionDenied:
		return "PermissionDenied"
	case CodeResourceExhausted:
		return "ResourceExhausted"
	case CodeFailedPrecondition:
		return "FailedPrecondition"
	case CodeAborted:
		return "Aborted"
	case CodeOutOfRange:
		return "OutOfRange"
	case CodeUnimplemented:
		return "Unimplemented"
	case CodeInternal:
		return "Internal"
	case CodeUnavailable:
		return "Unavailable"
	case CodeDataLoss:
		return "DataLoss"
	case CodeUnauthenticated:
		return "Unauthenticated"
	default:
		return "Unknown"
	}
}

File: trevrpc-rust/src/status.rs
use std::fmt;

use crate::{Metadata, RpcResponse};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Code {
    Ok,
    Cancelled,
    Unknown,
    InvalidArgument,
    DeadlineExceeded,
    NotFound,
    AlreadyExists,
    PermissionDenied,
    ResourceExhausted,
    FailedPrecondition,
    Aborted,
    OutOfRange,
    Unimplemented,
    Internal,
    Unavailable,
    DataLoss,
    Unauthenticated,
}

impl Code {
    /// Returns the wire-format numeric status code.
    #[must_use]
    pub const fn as_u32(self) -> u32 {
        match self {
            Self::Ok => 0,
            Self::Cancelled => 1,
            Self::Unknown => 2,
            Self::InvalidArgument => 3,
            Self::DeadlineExceeded => 4,
            Self::NotFound => 5,
            Self::AlreadyExists => 6,
            Self::PermissionDenied => 7,
            Self::ResourceExhausted => 8,
            Self::FailedPrecondition => 9,
            Self::Aborted => 10,
            Self::OutOfRange => 11,
            Self::Unimplemented => 12,
            Self::Internal => 13,
            Self::Unavailable => 14,
            Self::DataLoss => 15,
            Self::Unauthenticated => 16,
        }
    }

    /// Converts a wire-format numeric status code into a `Code`.
    #[must_use]
    pub const fn from_u32(code: u32) -> Self {
        match code {
            0 => Self::Ok,
            1 => Self::Cancelled,
            3 => Self::InvalidArgument,
            4 => Self::DeadlineExceeded,
            5 => Self::NotFound,
            6 => Self::AlreadyExists,
            7 => Self::PermissionDenied,
            8 => Self::ResourceExhausted,
            9 => Self::FailedPrecondition,
            10 => Self::Aborted,
            11 => Self::OutOfRange,
            12 => Self::Unimplemented,
            13 => Self::Internal,
            14 => Self::Unavailable,
            15 => Self::DataLoss,
            16 => Self::Unauthenticated,
            _ => Self::Unknown,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Status {
    code: Code,
    message: String,
    metadata: Metadata,
}

impl Status {
    /// Creates a status from a status code and message.
    #[must_use]
    pub fn new(code: Code, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            metadata: Metadata::new(),
        }
    }

    /// Creates an OK status.
    #[must_use]
    pub fn ok() -> Self {
        Self::new(Code::Ok, "")
    }

    /// Creates a cancelled status.
    #[must_use]
    pub fn cancelled(message: impl Into<String>) -> Self {
        Self::new(Code::Cancelled, message)
    }

    /// Creates an internal error status.
    #[must_use]
    pub fn internal(message: impl Into<String>) -> Self {
        Self::new(Code::Internal, message)
    }

    /// Creates an invalid-argument status.
    #[must_use]
    pub fn invalid_argument(message: impl Into<String>) -> Self {
        Self::new(Code::InvalidArgument, message)
    }

    /// Creates a deadline-exceeded status.
    #[must_use]
    pub fn deadline_exceeded(message: impl Into<String>) -> Self {
        Self::new(Code::DeadlineExceeded, message)
    }

    /// Creates a not-found status.
    #[must_use]
    pub fn not_found(message: impl Into<String>) -> Self {
        Self::new(Code::NotFound, message)
    }

    /// Creates a resource-exhausted status.
    #[must_use]
    pub fn resource_exhausted(message: impl Into<String>) -> Self {
        Self::new(Code::ResourceExhausted, message)
    }

    /// Creates an unavailable status.
    #[must_use]
    pub fn unavailable(message: impl Into<String>) -> Self {
        Self::new(Code::Unavailable, message)
    }

    /// Creates a failed-precondition status.
    #[must_use]
    pub fn failed_precondition(message: impl Into<String>) -> Self {
        Self::new(Code::FailedPrecondition, message)
    }

    /// Creates an unauthenticated status.
    #[must_use]
    pub fn unauthenticated(message: impl Into<String>) -> Self {
        Self::new(Code::Unauthenticated, message)
    }

    /// Creates an unimplemented status.
    #[must_use]
    pub fn unimplemented(message: impl Into<String>) -> Self {
        Self::new(Code::Unimplemented, message)
    }

    /// Returns the status code.
    #[must_use]
    pub const fn code(&self) -> Code {
        self.code
    }

    /// Returns the status message.
    #[must_use]
    pub fn message(&self) -> &str {
        &self.message
    }

    /// Returns terminal or response metadata carried by this status.
    #[must_use]
    pub fn metadata(&self) -> &Metadata {
        &self.metadata
    }

    /// Attaches metadata to this status.
    #[must_use]
    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
        self.metadata = metadata;
        self
    }

    /// Splits the status into its code and message.
    #[must_use]
    pub fn into_parts(self) -> (Code, String) {
        (self.code, self.message)
    }

    /// Returns whether this status is OK.
    #[must_use]
    pub const fn is_ok(&self) -> bool {
        matches!(self.code, Code::Ok)
    }

    /// Converts the status and response body into an RPC response.
    #[must_use]
    pub fn into_response(self, body: Vec<u8>) -> RpcResponse {
        let metadata = self.metadata.clone();
        self.into_response_with_metadata(body, metadata)
    }

    /// Converts the status, response body, and metadata into an RPC response.
    #[must_use]
    pub fn into_response_with_metadata(
        self,
        body: Vec<u8>,
        metadata: crate::Metadata,
    ) -> RpcResponse {
        RpcResponse {
            status: self.code.as_u32(),
            message: self.message,
            body,
            metadata,
        }
    }

    /// Builds a status from an RPC response status code, message, and metadata.
    #[must_use]
    pub fn from_response(response: &RpcResponse) -> Self {
        Self::new(Code::from_u32(response.status), response.message.clone())
            .with_metadata(response.metadata.clone())
    }
}

impl fmt::Display for Status {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        if self.message.is_empty() {
            write!(formatter, "{:?}", self.code)
        } else {
            write!(formatter, "{:?}: {}", self.code, self.message)
        }
    }
}

impl std::error::Error for Status {}
